exporter: support creating blobs with zstd compression

tonistiigi opened the pull request:
@ktock @fuweid 

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

ktock commented on util/compression/compression.go:
```diff
@@ -27,10 +27,18 @@ const (
 )
 
+const (
+	mediaTypeDockerSchema2LayerZstd = images.MediaTypeDockerSchema2Layer + ".zstd"
```
Do we need to support `application/vnd.docker.image.rootfs.diff.tar.zstd`? Maybe we can forcibly turn on `oci-mediatype=true` when zstd.

ktock replied:
IIUC no `application/vnd.docker.image.rootfs.diff.tar.zstd` in https://github.com/distribution/distribution/blob/v2.7.1/docs/spec/manifest-v2-2.md#media-types ?

tonistiigi replied:
I think this can provide more compatibility with registries. Not hard to support it and impossible for the spec to define something incompatible.

AkihiroSuda replied:
I'd prefer to keep this OCI-only.

I guess non-OCI compliant registries are unlikely to support `MediaTypeDockerSchema2Layer + ".zstd"` too?

tonistiigi replied:
Docker hub seems to support the docker mediatypes without any changes. I would guess at least Moby should always support them as well (after general zstd update).

tonistiigi replied:
fwiw, zstd is not part of any standard atm. Even OCI has never released it. I guess if `distribution/distribution` would ever add this mediatype (I'm not sure if they are interested in maintaining this part) it wouldn't be before docker engine actually supports it.

ktock reviewed (commented):
`TestGetRemote` in `manager_test.go` needs to be updated to also create zstd.

tonistiigi commented:
> TestGetRemote in manager_test.go needs to be updated to also create zstd.

I updated the test but found a very problematic issue. I never realized that estagz messes up the uncompressed digest for a reference. That is a huge problem as these digests are meant to be stable for the ref and image. `TestGetRemote` is supposed to check the digest stability and the only reason it passed was because there were a bunch of exceptions where if a timing caused a digest to change it just ignored the error and replaced the expected value from an internal array. I removed all the estargz parts from that test atm because it doesn't work and it is pointless to add more exceptions for new combinations where this can go wrong. We can discuss in a separate issue but I don't think we can ever support lossy compressions. If we want to keep estargz I think there at least needs to be a custom lossless decompress method. Even if that means that diffid and image config still changes at least we can use it for blob conversion. There can't be a case where build returns completely different images(with different diffid) because of the order of how a previous build (possibly build for different project) chose compressions.

ktock reviewed (approved):
Thank you for pointing the issue. I'll work on estargz decompressor that provides the original tar contents.

zstd changes LGTM.

fuweid reviewed (approved):
LGTM(nb)

AkihiroSuda commented on cache/manager_test.go:
```diff
@@ -1160,7 +1166,7 @@ func TestGetRemote(t *testing.T) {
 	for _, ir := range refs {
 		ir := ir.(*immutableRef)
-		for _, compressionType := range []compression.Type{compression.Uncompressed, compression.Gzip, compression.EStargz} {
+		for _, compressionType := range []compression.Type{compression.Uncompressed, compression.Gzip /*compression.EStargz,*/, compression.Zstd} {
```
Is EStargz intentionally omitted here? Could you add a comment line if so?

tonistiigi replied:
Yes, see https://github.com/moby/buildkit/pull/2344#issuecomment-913921170 and the commit message. Also have "disabled" message in code (not on that line). In current form, it is a release blocker for `v0.10` for me.
